fem/particleset.cpp

```diff
@@ -0,0 +1,902 @@
+               "Particle is incompatible with ParticleSet.");
+
+   // Add the particle
+   Array<int> idxs;
+   AddParticles(Array<int>({id_counter}), &idxs);
+   id_counter += id_stride;
+
+   // Set the new particle data
+   int idx = idxs[0];
+   SetParticle(idx, p);
+}
+
+void ParticleSet::AddParticles(int num_particles, Array<int> *new_indices)
+{
+   Array<int> add_ids(num_particles);
+   for (int i = 0; i < num_particles; i++)
+   {
+      add_ids[i] = id_counter;
+      id_counter += id_stride;
+   }
+
+   AddParticles(add_ids, new_indices);
+}
+
+void ParticleSet::RemoveParticles(const Array<int> &list)
+{
+   // Delete IDs
+   ids.DeleteAt(list);
+
+   // Delete data
+   for (int f = -1; f < GetNFields(); f++)
+   {
+      ParticleVector &pv = (f == -1 ? coords : *fields[f]);
+      pv.DeleteParticles(list);
+   }
+
+   // Delete tags
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      tags[t]->DeleteAt(list);
+   }
+}
+
+Particle ParticleSet::GetParticle(int i) const
+{
+   Particle p = CreateParticle();
+
+   Coords().GetValues(i, p.Coords());
+
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      Field(f).GetValues(i, p.Field(f));
+   }
+
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      p.Tag(t) = Tag(t)[i];
+   }
+
+   return p;
+}
+
+bool ParticleSet::ParticleRefValid() const
+{
+   if (coords.GetOrdering() == Ordering::byNODES)
+   {
+      return false;
+   }
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      if (fields[f]->GetOrdering() == Ordering::byNODES)
+      {
+         return false;
+      }
+   }
+   return true;
+}
+
+Particle ParticleSet::GetParticleRef(int i)
+{
+   Particle p = CreateParticle();
+
+   Coords().GetValuesRef(i, p.Coords());
+
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      Field(f).GetValuesRef(i, p.Field(f));
+   }
+
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      p.SetTagRef(t, &(*tags[t])[i]);
+   }
+
+   return p;
+}
+
+void ParticleSet::SetParticle(int i, const Particle &p)
+{
+   Coords().SetValues(i, p.Coords());
+
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      Field(f).SetValues(i, p.Field(f));
+   }
+
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      Tag(t)[i] = p.Tag(t);
+   }
+}
+
+void ParticleSet::PrintCSV(const char *fname, int precision)
+{
+   Array<int> all_field_idxs, all_tag_idxs;
+
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      all_field_idxs.Append(f);
+   }
+
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      all_tag_idxs.Append(t);
+   }
+
+   PrintCSV(fname, all_field_idxs, all_tag_idxs, precision);
+}
+
+void ParticleSet::PrintCSV(const char *fname, const Array<int> &field_idxs,
+                           const Array<int> &tag_idxs, int precision)
+{
+   std::stringstream ss_header;
+
+   // Configure header:
+   std::array<char, 3> ax = {'X', 'Y', 'Z'};
+
+   ss_header << "id" << ",";
+
+#ifdef MFEM_USE_MPI
+   ss_header << "rank" << ",";
+#endif // MFEM_USE_MPI
+
+   for (int f = -1; f < field_idxs.Size(); f++)
```

Comment:
Merging the loops for coords and fields does not make much sense here, it is just harder to read.